packages/@aws-cdk/cdk/lib/core/tokens.ts

```diff
@@ -133,3 +179,128 @@ export function resolve(obj: any, prefix?: string[]): any {
 
     return result;
 }
+
+import { FnConcat } from "../cloudformation/fn";
```

Comment:
The only thing I can think of that makes sense in order to decouple this from CloudFormation is to abstract this and pass as an argument to `resolve`. But this is totally fine for now I guess.

Reply:
I actually thought of an approach to decouple - the core idea is that a token should be expanded to `FnConcat` only if it's a CFN intrinsic function (or `Ref`). Otherwise, we don't really know what to do with it.

So basically the heuristics are if `resolve(x)` returns an object that looks like a CFN intrinsic (i.e. key starts with `Fn::` or `Ref`), then we can expand using `FnConcat`. Otherwise, we should probably fail resolution because that's probably not what the user wanted, unless someone can tell us how to expand this token.

So we can employ a "plug in" pattern - `resolve` can consult with a chain of "expanders" (?) when it needs to expand a token. And the CloudFormation library can register an expander that knows how to resolve `{Fn::}` and `Ref` objects. Some other libraries can use this to implement other expansion strategies.